src/mm/checkpoints.c

```diff
@@ -41,79 +41,198 @@
 #include <incremental_state_saving.h>
 
 
+void update_log_ckpt_statistics(int lid, double size, clock_timer time, bool is_incremental) {
 
-bool check_not_used_chunk_and_clean(malloc_area *m_area, int *bitmap_blocks) {
 
-	int j;
+	if (!is_incremental) {
+
+		statistics_post_lp_data(lid, STAT_CKPT_TIME, (double)clock_timer_value(time));
+		statistics_post_lp_data(lid, STAT_CKPT_MEM, size);
+	} else {
+
+		statistics_post_lp_data(lid, STAT_INC_CKPT, 1.0);
+		statistics_post_lp_data(lid, STAT_INC_CKPT_TIME, (double)clock_timer_value(time));
+		statistics_post_lp_data(lid, STAT_INC_CKPT_MEM, size);
+	}
+
+
+	
+}
+
+
+void update_restore_ckpt_statistics(int lid, clock_timer time, bool is_incremental) {
+
+	if (!is_incremental) {
+
+		statistics_post_lp_data(lid, STAT_RECOVERY_TIME, (double)clock_timer_value(time));
+	}else {
+		statistics_post_lp_data(lid, STAT_INC_RESTORE, 1.0);
+		statistics_post_lp_data(lid, STAT_INC_RECOVERY_TIME, (double)clock_timer_value(time));
+	}
+
+	
+}
+
+
+int compute_bitmap_blocks(int chunks) {
+
+	int blocks = chunks / NUM_CHUNKS_PER_BLOCK;
+	if (blocks < 1)
+		blocks = 1;
+
+	return blocks;
+
+}
+
+void set_chunk_dirty_from_address(void *addr, int lp) {
+
+	size_t chunk_size, area_size;
+	int j, blocks, num_chunks;
+	unsigned long long marea_base_addr;
+	malloc_area *target_marea;
+	malloc_state *m_state;
+	unsigned long long offset;
+	unsigned int index;
+
+	unsigned long long start_address = (unsigned long long)addr;
+
+	m_state = recoverable_state[lp];
+	int num_areas = m_state->num_areas;
+
+	malloc_area *areas = m_state->areas;
+	assert(areas != NULL);
+	
+
+	for (j=0; j < num_areas; j++) {
+		num_chunks = areas[j].num_chunks;
+		chunk_size = GET_AREA_CHUNK_SIZE(areas[j]);
+		blocks = compute_bitmap_blocks(num_chunks);
+		area_size = num_chunks * chunk_size;
+		marea_base_addr = ((unsigned long long) areas[j].area); 
+		if (start_address >= marea_base_addr && start_address < marea_base_addr + area_size) {
+			target_marea = &areas[j];
+			offset = start_address - marea_base_addr;
+			index = offset / chunk_size;
+			SET_DIRTY_BIT(target_marea, index); 
+			target_marea->dirty_chunks++;
+			
+			if (target_marea->is_recoverable) {
+				if(target_marea->state_changed == 0) {
+					m_state->dirty_bitmap_size += blocks * BLOCK_SIZE;
+					m_state->dirty_areas++;
+					target_marea->last_access = current_lvt;
+					m_state->total_inc_size += chunk_size;
+				}
```

Comment:
These is something strange here.
only the bitmap size should be added conditionally to state_changed values
```suggestion
				if(target_marea->state_changed == 0) {
					m_state->dirty_bitmap_size += blocks * BLOCK_SIZE;
					m_state->dirty_areas++;
				}
					target_marea->last_access = current_lvt;
					m_state->total_inc_size += chunk_size;
```

Reply:
correct